### ISSUE TITLE
'XML' Package Error

### ISSUE THREAD

**Issue by mdbarnesUCSD** [sagar87/tensorsignatures#6]
Hello,

I am following the guide on the README for running tensorsignatures. I installed by using ‘pip install tensorsignatures’ and am using R-3.4.4.

I ran the two lines in the ‘preparing input data using a custom installation’ section:
```
Rscript -e "source('https://bioconductor.org/biocLite.R'); biocLite('VariantAnnotation')"
Rscript -e "source('https://bioconductor.org/biocLite.R'); biocLite('rhdf5')"
```

Then afterwards, I ran:
`Rscript processVcf.R example1.vcf example2.vcf example_output.h5`

And got this error message:
```
Error: package or namespace load failed for ‘VariantAnnotation’ in loadNamespace(i, c(lib.loc, .libPaths()), versionCheck = vI[[i]]):
 there is no package called ‘XML’
```

So, then I ran:
```
Install.packages(“XML”)
Warning message:
package ‘XML’ is not available (for R version 3.4.4)
```

I see that the current version of ‘XML’ is for R>=4.0.0.

Could you please let me know which version of R and XML works?

Thank you!

**Comment by sagar87** [comment]
Hello mdbarnesUCSD,

thanks for your interest in `tensorsignatures`. Judging from the error message, it seems that the bioconductor package `VariantAnnotation` was not installed properly. Unfortunately, bioconductor dependencies seem to break every time a new version of R is rolled out, I myself also had a lot of problems with that. I will look into that and update the install instructions for the latest version of  R. 

In the mean time, you could try to download and run the docker image which contains R + all required dependencies (you can find the instructions on the front page of the repository).

Let me know if that works for you.

**Comment by mdbarnesUCSD** [comment]
Thank you for your response. I am trying the docker image and am hoping you could clarify the workflow for me.

These are the steps I have taken for working with the docker image:

```
$ git clone https://github.com/gerstung-lab/tensorsignatures.git
$ cd tensorsignatures
$ docker-compose up --build
```

After having run the commands above, I went to look at the notebooks with tutorials at:
http://localhost:8889.

The `http://localhost:8889.` leads me to a login page, and I used the authentication token generated from `docker-compose up --build` to connect to the server but I do not see any tutorials or documentation here.

I tried to run 'Rscript processVcf.R <vcf1.vcf> <vcf2.vcf> ... <vcfn.vcf> <output.h5>' locally by doing the following from the directory with my input files:

```
$ docker pull sagar87/tensorsignatures-data:latest
$ docker run -v $PWD:/Users/mbarnes/Desktop/test_tensor sagar87/tensorsignatures-data example1.vcf example2.vcf example_output.h5
```

But I am receiving the following errors:

Bioconductor version 3.12 (BiocManager 1.30.10), ?BiocManager::install for help
[1] "Loading vcf(s) ..."
Error in h(simpleError(msg, call)) : 
  error in evaluating the argument 'x' in selecting a method for function 'seqinfo': file(s) do not exist:
  ‘example1.vcf’
Calls: lapply ... <Anonymous> -> open -> open.BcfFile -> .io_check_exists
Execution halted

Please let me know if there is any more information I can provide.

**Comment by sagar87** [comment]
Are you sure you pulled the latest version of the repository ? I just tried to pull and start the docker image and it seemed to work fine. In any case, jupyter notebooks should be located in the [mount folder](https://github.com/gerstung-lab/tensorsignatures/tree/master/mount) of the repository, would you mind trying to pull the repository again ? Another possibility is to download the repository as a zip file, unpack it and run `docker-compose`, both approaches should work. 
![Image 29-08-2020 at 06 07](https://user-images.githubusercontent.com/7542594/91629130-198dc180-e9be-11ea-90a7-cb945b8ddde3.jpg)
BTW: you are probably one of the first persons to read the documentation, presumably there is a lot room for improvement, so feedback is appreciated ;). 

Assuming you are in the same folder where `example1.vcf` and `example2.vcf` are, could you try to run the following command (simply exchange `/Users/mbarnes/Desktop/test_tensor` with `/usr/src/app`):

`$ docker run -v $PWD:/usr/src/app sagar87/tensorsignatures-data  example1.vcf example2.vcf example_output.h5`

Basically what the `-v` flag does is to mount your current folder (`$PWD`) into the folder `/usr/src/app` of the docker image (this is where all the scripts and annotation data is located).

So if `example1.vcf` and `example2.vcf` are in `/Users/mbarnes/Desktop/test_tensor` you could also run

`$ docker run -v /Users/mbarnes/Desktop/test_tensor:/usr/src/app sagar87/tensorsignatures-data  example1.vcf example2.vcf example_output.h5`

which writes `example_output.h5` into `/Users/mbarnes/Desktop/test_tensor`.

**Comment by mdbarnesUCSD** [comment]
Thanks, I did the following:

git clone https://github.com/gerstung-lab/tensorsignatures
cd tensorsignatures
docker-compose up --build

After signing into the Jupyter notebook I see the four tutorials.

Next, I tried to do the 'preparing input data using docker' section:
docker pull sagar87/tensorsignatures-data:latest
docker run -v /Users/mbarnes/Desktop/test_tensor/input:/usr/src/app sagar87/tensorsignatures-data example1.vcf example2.vcf output_test.h5

Which produced this error:
docker: Error response from daemon: OCI runtime create failed: container_linux.go:349: starting container process caused "exec: \"/usr/src/app/entrypoint.sh\": stat /usr/src/app/entrypoint.sh: no such file or directory": unknown.
ERRO[0000] error waiting for container: context canceled 

I checked and do not have a /usr/src/app directory. Does docker create this directory or do I need to change the path in the -v parameter?

**Comment by sagar87** [comment]
Hi mdbarnesUCSD,

apologies, I looked into this and found the bug. I uploaded a new image and updated the README. Could you try to pull the image again 

`docker pull sagar87/tensorsignatures-data:latest`

and then run the image like this

`docker run -v /Users/mbarnes/Desktop/test_tensor/input:/usr/src/app/mount sagar87/tensorsignatures-data example1.vcf example2.vcf output_test.h5`.

The problem was that I was testing the image only from my built folder which contained the `vcf` files as well as the scripts and genome data. However, when mounting any other folder into the container, docker replaces all files in `/usr/src/app` with the content of the mounted folder, and for this reason `entrypoint.sh` was missing. I fixed this by adding mount folder to the docker image (such that the content of `/usr/src/app` does not get overridden). It should work now. Sorry about that, and thanks for your patience.

**Comment by mdbarnesUCSD** [comment]
Okay, I have pulled the new image and have run the code:
`docker run -v /Users/mbarnes/Desktop/test_tensor/input:/usr/src/app/mount sagar87/tensorsignatures-data example1.vcf example2.vcf output_test.h5`

And received this error:
Bioconductor version 3.12 (BiocManager 1.30.10), ?BiocManager::install for help
/usr/src/app/entrypoint.sh: line 2:     7 Killed                  Rscript /usr/src/app/process.R "$@"

Thanks again for your help.

**Comment by sagar87** [comment]
We're getting there XD. I think this error results from allocating too less memory for docker. If you're on the Mac you can change the settings in the Docker Desktop application (by default docker reserves only 2GB I think, which is way too less for R with bioconductor).
[Screenshot 2020-09-01 at 19.00.07.pdf](https://github.com/sagar87/tensorsignatures/files/5158009/Screenshot.2020-09-01.at.19.00.07.pdf)

**Comment by mdbarnesUCSD** [comment]
Awesome, I updated the default Docker settings and the data preparation section worked.

When I provided an input vcf file without a header I was receiving this error:
Bioconductor version 3.12 (BiocManager 1.30.10), ?BiocManager::install for help
[1] "Loading vcf(s) ..."
[E::COMPAT_bcf_hdr_read] Input is not detected as bcf or vcf format
Error in h(simpleError(msg, call)) : 
  error in evaluating the argument 'x' in selecting a method for function 'seqinfo': no 'header' line "#CHROM POS ID..."?
Calls: lapply ... scanBcfHeader -> scanBcfHeader -> .handleSimpleError -> h

Is the VCF header used for any information other than the columns for CHROM, POS, ID, REF, and ALT?

Also, after producing the tsdata.h5 (with a file w/ a header section) file I am attempting to run the "prep" command:
`tensorsignatures prep input_f/output_test.h5 input_f/tsdata.h5`

But received this error:
/Users/mbarnes/anaconda3/envs/tensor_aug31/lib/python3.6/site-packages/tensorsignatures-0.5.0-py3.6.egg/tensorsignatures/util.py:978: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  col1 = snv[[slice(None)] * (snv.ndim - 3) + [0] + [slice(None)] * 2]
/Users/mbarnes/anaconda3/envs/tensor_aug31/lib/python3.6/site-packages/tensorsignatures-0.5.0-py3.6.egg/tensorsignatures/util.py:992: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  col2.append(snv[idx])

Does the Docker image support functions for steps 2 and 3?

Thanks!

**Comment by sagar87** [comment]
Regarding the vcf files, they must be parseable with the [readVcf](https://www.rdocumentation.org/packages/VariantAnnotation/versions/1.18.5/topics/readVcf) function of the variant annotation package.

The second command should have worked. Can you check whether the output file exists, i.e. `input_f/ts.data.h5` was   written to the disk ?  For now you can ignore any `Warning` messages (you'll see more of them).

**Comment by mdbarnesUCSD** [comment]
Great, thank you! 

You are right, the "errors" were actually just warnings and `input_f/ts.data.h5` was written to the disk.

**Comment by sagar87** [comment]
Great, thanks for confirming.
